File: backend/routes/appointments.py
"""
routes/appointments.py — Appointment booking, management, and email notifications
Mirrors backend/routes/appointments.js
"""
from fastapi import APIRouter, Depends, HTTPException, BackgroundTasks
from pydantic import BaseModel
from typing import Optional, Any, Dict

from middleware.auth import get_current_user
from models.appointment import Appointment
from models.doctor import Doctor
from models.user import User
from models.hospital import Hospital
from services.email_service import send_appointment_confirmation_email

router = APIRouter(prefix="/api/appointments", tags=["Appointments"])


class AppointmentCreate(BaseModel):
    doctorId: Optional[str] = None
    hospitalId: Optional[str] = None
    date: Optional[str] = None
    timeSlot: Optional[Any] = None
    department: Optional[str] = None
    reason: Optional[str] = None


class StatusUpdate(BaseModel):
    status: str


class PaymentUpdate(BaseModel):
    paymentProof: str


async def _send_confirmation_email(user: Dict[str, Any], apt: Dict[str, Any], doctor_id: Optional[str], hospital_id: Optional[str]):
    try:
        doctor_name = "Assigned Doctor"
        hospital_name = ""
        if doctor_id:
            doc = Doctor.find_by_id(doctor_id)
            if doc and doc.get("user_id"):
                doc_user = User.find_by_id(doc["user_id"])
                if doc_user:
                    doctor_name = f"Dr. {doc_user.get('name')}"
        if hospital_id:
            hosp = Hospital.find_by_id(hospital_id)
            if hosp:
                hospital_name = hosp.get("name", "")

        send_appointment_confirmation_email(
            to=user.get("email"),
            name=user.get("name"),
            appointment=apt,
            doctor_name=doctor_name,
            hospital_name=hospital_name
        )
    except Exception as e:
        print(f"[Appointments] Email confirmation failed: {e}")


@router.post("")
@router.post("/")
async def create_appointment(payload: AppointmentCreate, background_tasks: BackgroundTasks, user: dict = Depends(get_current_user)):
    try:
        apt_data = {
            "patientId": user["id"],
            "patientName": user.get("name"),
            "patientEmail": user.get("email"),
            "patientPhone": user.get("phone"),
            "doctorId": payload.doctorId,
            "hospitalId": payload.hospitalId,
            "date": payload.date,
            "timeSlot": payload.timeSlot,
            "department": payload.department,
            "reason": payload.reason,
        }
        appointment = Appointment.create(apt_data)

        background_tasks.add_task(
            _send_confirmation_email,
            user=user,
            apt=appointment,
            doctor_id=payload.doctorId,
            hospital_id=payload.hospitalId
        )

        return {"success": True, "appointment": appointment}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/my")
async def get_my_appointments(user: dict = Depends(get_current_user)):
    try:
        appointments = Appointment.find_by_patient(user["id"])
        enriched = []
        for apt in appointments:
            doc_id = apt.get("doctor_id")
            hosp_id = apt.get("hospital_id")

            doctor = Doctor.find_by_id(doc_id) if doc_id else None
            hospital = Hospital.find_by_id(hosp_id) if hosp_id else None

            doctor_user = None
            if doctor and doctor.get("user_id"):
                doctor_user = User.find_by_id(doctor["user_id"])

            enriched.append({
                **apt,
                "patient": {
                    "id": apt.get("patient_id"),
                    "name": apt.get("patient_name"),
                    "email": apt.get("patient_email"),
                    "phone": apt.get("patient_phone"),
                },
                "doctor": {
                    **doctor,
                    "userId": {
                        "id": doctor_user["id"],
                        "name": doctor_user.get("name"),
                        "email": doctor_user.get("email"),
                    } if doctor_user else None
                } if doctor else None,
                "hospital": {
                    "id": hospital.get("id"),
                    "name": hospital.get("name"),
                    "address": hospital.get("address"),
                } if hospital else None,
            })
        return {"success": True, "appointments": enriched}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/all")
async def get_all_appointments(user: dict = Depends(get_current_user)):
    if user.get("role") != "admin":
        raise HTTPException(status_code=403, detail="Admin access required.")
    try:
        appointments = Appointment.find_all()
        return {"success": True, "appointments": appointments}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.put("/{apt_id}/status")
async def update_appointment_status(apt_id: str, payload: StatusUpdate, user: dict = Depends(get_current_user)):
    try:
        appointment = Appointment.update_status(apt_id, payload.status)
        return {"success": True, "appointment": appointment}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.put("/{apt_id}/payment")
async def update_appointment_payment(apt_id: str, payload: PaymentUpdate, user: dict = Depends(get_current_user)):
    try:
        appointment = Appointment.update_payment(apt_id, payload.paymentProof)
        return {"success": True, "appointment": appointment}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: backend/routes/appointments.js
const express = require('express');
const router = express.Router();
const { auth } = require('../middleware/auth');
const Appointment = require('../models/Appointment');
const Doctor = require('../models/Doctor');
const User = require('../models/User');
const Hospital = require('../models/Hospital');
const { sendAppointmentConfirmationEmail } = require('../services/emailService');

// Book appointment
router.post('/', auth, async (req, res) => {
  try {
    const { doctorId, hospitalId, date, timeSlot, department, reason } = req.body;
    const user = req.user;

    const appointment = await Appointment.create({
      patientId: user.id,
      patientName: user.name,
      patientEmail: user.email,
      patientPhone: user.phone,
      doctorId,
      hospitalId,
      date,
      timeSlot,
      department,
      reason,
    });

    // Resolve doctor & hospital names for the email (non-blocking)
    (async () => {
      try {
        let doctorName = 'Assigned Doctor';
        let hospitalName = '';
        if (doctorId) {
          const doc = await Doctor.findById(doctorId);
          if (doc) {
            const docUser = await User.findById(doc.user_id);
            if (docUser) doctorName = `Dr. ${docUser.name}`;
          }
        }
        if (hospitalId) {
          const hosp = await Hospital.findById(hospitalId);
          if (hosp) hospitalName = hosp.name;
        }
        await sendAppointmentConfirmationEmail({
          to: user.email,
          name: user.name,
          appointment,
          doctorName,
          hospitalName,
        });
      } catch (err) {
        console.error('Appointment email failed:', err.message);
      }
    })();

    res.status(201).json({ success: true, appointment });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
});

// Get user appointments
router.get('/my', auth, async (req, res) => {
  try {
    const appointments = await Appointment.findByPatient(req.userId);

    // Manually enrich with doctor and hospital info
    const enriched = await Promise.all(
      appointments.map(async (apt) => {
        const [doctor, hospital] = await Promise.all([
          Doctor.findById(apt.doctor_id),
          apt.hospital_id ? Hospital.findById(apt.hospital_id) : null,
        ]);

        let doctorUser = null;
        if (doctor) {
          doctorUser = await User.findById(doctor.user_id);
        }

        return {
          ...apt,
          patient: { id: apt.patient_id, name: apt.patient_name, email: apt.patient_email, phone: apt.patient_phone },
          doctor: doctor ? { ...doctor, userId: doctorUser ? { id: doctorUser.id, name: doctorUser.name, email: doctorUser.email } : null } : null,
          hospital: hospital ? { id: hospital.id, name: hospital.name, address: hospital.address } : null,
        };
      })
    );

    res.json({ success: true, appointments: enriched });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
});

// Get all appointments (admin) - shows who booked with their email
router.get('/all', auth, async (req, res) => {
  try {
    if (req.user.role !== 'admin') {
      return res.status(403).json({ success: false, message: 'Admin access required.' });
    }
    const appointments = await Appointment.findAll();
    res.json({ success: true, appointments });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
});

// Update appointment status
router.put('/:id/status', auth, async (req, res) => {
  try {
    const appointment = await Appointment.updateStatus(req.params.id, req.body.status);
    res.json({ success: true, appointment });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
});

// Upload payment proof
router.put('/:id/payment', auth, async (req, res) => {
  try {
    const appointment = await Appointment.updatePayment(req.params.id, req.body.paymentProof);
    res.json({ success: true, appointment });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
});

module.exports = router;
